lib/compiler/passes/report-incorrect-plucking.js

```diff
@@ -18,9 +18,7 @@ function reportIncorrectPlucking(ast) {
       if (node.pick) {
         if (action) {
           throw new GrammarError(
-            "\"@\" cannot be used with an action block "
-              + "at line " + node.location.start.line + ", "
-              + "column " + node.location.start.column + ".",
```

Comment:
You should replace self location with action node location, but there is a subtle aspect. `action` node `location` points to the expression and associated code block and it's `start` position visually not related to the code block at all. We have two options here, how to solve that:
- change `location` of the action node. I think that is bad, because:
  - that is breaking change
  - that breaks assumption that locations of all children nodes lives in the location of their parent node
- introduce a new `codeLocation` property, which should hold the `CodeBlock` location, i.e. space from opening `{` till closing `}`. That properties should be added to all nodes which have a `CodeBlock`. For me, that is the best choice

After introducing that property just use `action.codeLocation.start` there

Reply:
Also, `codeLocation` would used for source map support

Reply:
why doesn't the code block have it's own AST node?

Reply:
Actually, it have -- that is `initializer`, `action`, `semantic_and`, and `semantic_not`

Reply:
I'm suggesting that the ast should be:

```js
{
  type: 'semantic_and',
  code: {
    type: 'code',
    source: ' return true; '
    location: ...
  },
  location: ...
}
```

instead of:

```js
{
  type: 'semantic_and',
  code: ' return true; ',
  location: ...
}
```

Reply:
I think, it would be hard to process them if use the visitor pattern as with other kinds of nodes. You always need information from their parent node to process it. So you have to:
- or give each `code` node different name -- but then this situation wouldn't differ from present
- or not use visitor pattern when process such nodes -- then an extra entity will give nothing useful
- or give a reference to the owner node down to the visitor so `code` nodes known which data their represents

I think, that all alternatives not gives us so much and just a new property with an extra location will be enough:
```js
{
  type: 'semantic_and',
  code: ' return true; ',
  codeLocation: ..., // for initializer just the same reference, as location
  location: ...
}
```